### PR TITLE
aggregation: "group_concat" should not modify the argument during execution

### LINKED ISSUE #5663: tidb-server panicked while executing some sql with plan-cache enabled
1. What did you do?

```toml
# config file:  test.toml
[plan-cache]
enabled = true
capacity = 2560
shards = 256
```

```bash
#start tidb-server with this config file:
bin/tidb-server --config test.toml
```

```sql
use test;
drop table t1;
create table t1 (i int unsigned, primary key(i));
insert into t1 values (1),(2),(3);
select group_concat(i) from t1 where i > 1;    
select group_concat(i) from t1 where i > 1; 
```

2. What did you expect to see?
should not panic

3. What did you see instead?

first select OK, second select panicked.

```
2018/01/17 10:27:24.097 conn.go:410: [error] lastCmd select group_concat(i) from t1 where i > 1, runtime error: index out of range, goroutine 171 [running]:
github.com/pingcap/tidb/server.(*clientConn).Run.func1(0xc420650340, 0xc42011de0f)
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/server/conn.go:408 +0x10e
panic(0x1cdbe20, 0x2867530)
        /Users/darren/.local/go1.9.2/src/runtime/panic.go:491 +0x283
github.com/pingcap/tidb/util/chunk.Row.GetString(...)
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/util/chunk/chunk.go:513
github.com/pingcap/tidb/util/chunk.(*Row).GetString(0xc420196dd0, 0x0, 0x147e600, 0x13ea493)
        <autogenerated>:1 +0x126
github.com/pingcap/tidb/expression.(*Column).EvalString(0xc420359c70, 0x2672940, 0xc420196dd0, 0xc4203436e0, 0x1043447, 0x40, 0x1dcf4c0, 0x1004401, 0xc420051c40)
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/expression/column.go:242 +0x30e
github.com/pingcap/tidb/expression/aggregation.(*concatFunction).initSeparator(0xc420304540, 0xc4203436e0, 0x2672940, 0xc420196dd0, 0x0, 0x1)
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/expression/aggregation/concat.go:41 +0x73
github.com/pingcap/tidb/expression/aggregation.(*concatFunction).Update(0xc420304540, 0xc420343aa0, 0xc4203436e0, 0x2672940, 0xc420196dd0, 0x0, 0xc4202d601c)
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/expression/aggregation/concat.go:57 +0x630
github.com/pingcap/tidb/executor.(*StreamAggExec).consumeOneGroup(0xc4200aa640, 0x2ca8810, 0xc4205a8f80, 0xc4203046c0, 0x0, 0x1)
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/executor/aggregate.go:356 +0xfb
github.com/pingcap/tidb/executor.(*StreamAggExec).NextChunk(0xc4200aa640, 0x2ca8810, 0xc4205a8f80, 0xc4203046c0, 0x1, 0x1)
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/executor/aggregate.go:333 +0x84
github.com/pingcap/tidb/executor.(*recordSet).NextChunk(0xc420343b00, 0x2ca8810, 0xc4205a8f80, 0xc4203046c0, 0xc420343b00, 0xc4203046c0)
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/executor/adapter.go:103 +0x56
github.com/pingcap/tidb/server.(*tidbResultSet).NextChunk(0xc4202dfe30, 0x2ca8810, 0xc4205a8f80, 0xc4203046c0, 0x4, 0x400)
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/server/driver_tidb.go:307 +0x51
github.com/pingcap/tidb/server.(*clientConn).writeChunks(0xc420650340, 0x2ca8810, 0xc4205a8f80, 0x2659c00, 0xc4202dfe30, 0x0, 0xc4202dfe30, 0x1)
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/server/conn.go:936 +0x3ef
github.com/pingcap/tidb/server.(*clientConn).writeResultset(0xc420650340, 0x2ca8810, 0xc4205a8f80, 0x2659c00, 0xc4202dfe30, 0xc4202c0000, 0x0, 0x0)
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/server/conn.go:861 +0x668
github.com/pingcap/tidb/server.(*clientConn).handleQuery(0xc420650340, 0x2ca8810, 0xc4205a8f80, 0xc4202d4bd1, 0x2a, 0x0, 0x0)
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/server/conn.go:810 +0x115
github.com/pingcap/tidb/server.(*clientConn).dispatch(0xc420650340, 0xc4202d4bd1, 0x2b, 0x2b, 0x0, 0x0)
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/server/conn.go:592 +0x595
github.com/pingcap/tidb/server.(*clientConn).Run(0xc420650340)
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/server/conn.go:452 +0x1b8
github.com/pingcap/tidb/server.(*Server).onConn(0xc420343c80, 0x265c900, 0xc4202be6f8)
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/server/server.go:316 +0x144
created by github.com/pingcap/tidb/server.(*Server).Run
        /Users/darren/Sources/go/src/github.com/pingcap/tidb/server/server.go:262 +0x496

2018/01/17 10:27:24.097 server.go:298: [info] [1] close connection
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

commit: 90d653c8caaee4119844b562a69c93ccf1a2db2a
```
Release Version: v1.1.0-alpha-444-g90d653c8
Git Commit Hash: 90d653c8caaee4119844b562a69c93ccf1a2db2a
Git Branch: master
UTC Build Time: 2018-01-17 02:00:15
```

### PR DESCRIPTION
**Pull request by zz-jason**
fix #5663

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
LGTM
/run-all-tests

File: expression/aggregation/concat.go [hunk 1]
```diff
@@ -60,8 +59,9 @@ func (cf *concatFunction) Update(ctx *AggEvaluateContext, sc *stmtctx.StatementC
 		}
 		cf.sepInited = true
 	}
-	for _, a := range cf.Args {
-		value, err := a.Eval(row)
+
+	for i, length := 0, len(cg.Args)-1; i < length; i++ {
```
Comment (winkyao) [inline expression/aggregation/concat.go]: add some descrition here, why len(cg.Args)-1

**Comment by shenli** [comment]
Please add a unit test case.

**Comment by winkyao** [comment]
@zz-jason Plz fix ci

**Comment by shenli** [comment]
LGTM

**Pull request merged**